src/FakeItEasy/DummyFactory.cs

```diff
@@ -37,7 +37,8 @@ public object Create(Type type)
         {
             this.AssertThatFakeIsOfCorrectType(type);
 
-            return this.Create();
+            return this.Create(); // this might be null. Same with any implementation of IDummyFactory.
+            // We could allow it and trap in DynamicDummyFactory, returning false there.

```

Comment:
I don't know... Can a dummy be null?

Reply:
> I generally don't _like_ the idea, but if users think `null` is an appropriate value to return, should we stop them?

I don't think so. There's no technical reason to forbid it. If users want to do that, and it hurts them later... I guess it's their problem?

> If we officially allow null Dummies, should we make sure we don't consider them as usable arguments for constructors where the parameter is a non-nullable reference type?

I'd say no. Non-nullable reference types is a compile time feature. There's no expectation that the constraint should be enforced at runtime. (and even at compile time, it's only a warning, not an error)

Reply:
> Likewise, if we say that null Dummies are allowed, should our own Dummy resolution algorithm fall back to returning a null when all other creation attempts fail? And if so, should it only do so when trying to generate an argument for a constructor, or should even `A.Dummy<T>` return null?

That's a harder question. I'd say keep the current behavior (i.e. never return a null dummy), but I'm no sure...

Reply:
> never return a null dummy

… unless a DummyFactory made it?

Yeah, I don't want to start returning nulls willy-nilly. Perhaps we can update the docs (along with whatever code changes to ensure we don't warn on null Dummies coming from DummyFactories) to

> FakeItEasy's own Dummy-creation strategies will never produce a null Dummy, but user-supplied DummyFactories may do so 

And leave it at that.

Reply:
> … unless a DummyFactory made it?

Exactly